File: cpp/94_BinaryTreeInOrderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
// in-order: left - root - right
class Solution {
public:
	vector<int> inorderTraversal(TreeNode* root) {
		vector<int> res;
		if (root == NULL) return res;
		stack<TreeNode*> stack;

		TreeNode* p_current = root;
		while (!stack.empty() || p_current) {
			if (p_current) {
				stack.push(p_current);
				p_current = p_current->left;
			}
			else {
				p_current = stack.top();
				stack.pop();
				res.push_back(p_current->val);
				p_current = p_current->right;
			}
		}
		return res;
	}
};

// recursion version
class Solution {
private:
	void recursion(TreeNode* node, vector<int> &res) {
		if (!node) return;
		recursion(node->left, res);
		res.push_back(node->val);
		recursion(node->right, res);
	}
public:
	vector<int> inorderTraversal(TreeNode* root) {
		vector<int> res;
		recursion(root, res);
		return res;
	}

};

File: java/94.binary-tree-inorder-traversal.java
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode id=94 lang=java
 *
 * [94] Binary Tree Inorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {
    List<Integer> res = new ArrayList<>();

    public List<Integer> inorderTraversal(TreeNode root) {
        if (root == null)
            return res;
        inorderTraversal(root.left);
        res.add(root.val);
        inorderTraversal(root.right);
        return res;
    }

    // iterative
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> ret = new ArrayList<>();
        Deque<TreeNode> stack = new LinkedList();
        TreeNode curr = root;
        while (curr != null || !stack.isEmpty()) {
            while (curr != null) {
                // keep push left children
                stack.push(curr);
                curr = curr.left;
            }
            // add parent
            curr = stack.pop();
            ret.add(curr.val);
            // go to right side
            curr = curr.right;
        }
        return ret;
    }
}
// @lc code=end
